keps/sig-node/2681-pod-host-ip/README.md

```diff
@@ -0,0 +1,555 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal aims to improve the Pod's ability to obtain the address of the node
+
+## Motivation
+
+### Goals
+
+- Field `status.hostIPs` added for Pod
+- Downward API support for `status.hostIPs`
+
+### Non-Goals
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Pod wants to know the IPv6 address of the node it is on
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Versioned API Change: PodStatus v1 core
+
+In order to maintain backwards compatibility for the core V1 API, this proposal
+retains the existing (singular) "HostIP" field in the core V1 version of the
+PodStatus V1 core API and adds a new array of structures that store host IPs
+along with associated metadata for that IP.
+
+``` go
+    // HostIP address of the host to which the pod is assigned. Empty if not yet scheduled.
+    // +optional
+    HostIP string
+
+    // HostIP represents the IP address of a host.
+    // IP address information. Each entry includes:
+    //    IP: An IP address allocated to the host.
+    type HostIP struct {
+      // ip is an IP address (IPv4 or IPv6) assigned to the host
+      IP string
+    }
```

Comment:
It's not clear to me that need a struct type here... and if we do, then it should probably just use `v1.NodeAddress`, and say that kubelet will copy the whole NodeAddress object for the primary and secondary IPs

Reply:
This is a reference to PodIPs, I think it’s best to keep it consistent with here.

Reply:
Being consistent with PodIPs is a good approach in one way, but that doesn't mean it's the best approach in every way.

Kubelet sets `pod.status.hostIP[s]` from `node.status.addresses`, which already has its own struct with metadata (`v1.NodeAddress`). It would be very weird to use some _other_ type for `hostIPs`, given that kubelet doesn't have any more information about the host's IPs other than what's in `node.status.addresses`, so if `hostIPs` was going to be a different type from `addresses`, it would necessarily have to be a subset of the same information.

OTOH, it's not clear that anyone would actually benefit from having `HostIPs []NodeAddress` rather than `HostIPs []string`. You've already said that the additional metadata won't show up in the downward API (just like with `podIPs`). So the benefit of having extra metadata here would only be to people retrieving Pods from the apiserver. But then they could just retrieve the pod's Node from the apiserver too if they needed additional information about the pod's host IPs (or if they needed information about host IPs other than the ones kubelet assigned to the pod...).

Reply:
I got it. 

I looked at the definition of `v1.NodeAddress`, and it seems that only `v1.NodeExternalIP` and `v1.NodeInternalIP` fits the meaning of HostIP in the Pod, It is still necessary to define a new struct type. 
``` go
const (
	NodeHostName    NodeAddressType = "Hostname"
	NodeExternalIP  NodeAddressType = "ExternalIP"
	NodeInternalIP  NodeAddressType = "InternalIP"
	NodeExternalDNS NodeAddressType = "ExternalDNS"
	NodeInternalDNS NodeAddressType = "InternalDNS"
)
```
https://github.com/kubernetes/kubernetes/blob/c3b7d3056a76c3acef087af79cf973f97817392e/staging/src/k8s.io/api/core/v1/types.go#L4931-L4946

In the actual function to obtain HostIP, the difference between `v1.NodeExternalIP` and `v1.NodeInternalIP` is ignored, but `v1.NodeInternalIP` is preferred by default
https://github.com/kubernetes/kubernetes/blob/c3b7d3056a76c3acef087af79cf973f97817392e/pkg/util/node/node.go#L96-L135

Reply:
singular value is a string
```
HostIP string
```

if we pluralise I  think we have to maintain the same type, https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md 
It seems we have to be able to populate the singular from the plural, and it looks like converting types is a breaking change, so I think it should be

```
HostIPs []string
```

we already have a method to get the HostIPs
https://github.com/kubernetes/kubernetes/blob/347730b6b49849a3b8501a2b09caf4aec738c9e4/pkg/kubelet/kubelet_getters.go#L265-L272 , I assume that we'll use that to get the value

Reply:
PING @danwinship

Reply:
We *could* add metadata, I just don't think we need to, for the reasons @danwinship cited.  Unlike PodIP (where metadata would be unique), this is just a copy of other data.

Reply:
I agree with dan initial comment, we should use NodeAddressType as a metadata with possibly additional properties. Every time i think of a use case where a pod (or other pods) needs hostIPs then it is probably in a context that would need a specific type of address (e.g. internal v. external).

I also imagine a situation where node has multiple IP but what is relevant to user is a subset of it. e.g. a node may have multiple internal addresses but one or a few of them are relevant (e.g. marked with certain properties). In this case either a modified node controller or additional control patches the node object. FWIW if we decided on using properties then let us put validation/limits on them early on. I remember discussing PodIP (struct) it had properties on them early on for similar reasons (e.g. pod with n IPs one of them is connected to BE network over vpn). But we had to remove because we couldn't come up with validation/limits for them.

@aojea I understand the guideline but unlike PodIPs this is fundamentally a new field with new semantics. btw PodIP is also string but PodIP is a slice of structs.

also having a slice of structs makes this less painful to further future rev (adding/removing fields) vs a slice of strings.

Reply:
> @aojea I understand the guideline but unlike PodIPs this is fundamentally a new field with new semantics. btw PodIP is also string but PodIP is a slice of structs.

agree with your reasoning, disagree with adding a new field and semantics to HostIPs :).

I think that there should be only one IP per family, the others IPs and its metadata are just implementation details that shouldn't be part of Kubernetes, the user/admin/... only need that the HostIPs are reachable, he can use routing, bridging, 1, 2 or 20 ips, is up to him how to do that.

The node object is a good example of how bad is having multiple IPs and metadata fields, I personally prefer to not leak networking infra into the API, it is a slippery slope ...

Reply:
NodeAddress is set via cloud provider or kubelet flag right? So user has control over what can `leak` into Pod.Status.HostIPs.